### PR TITLE
[GSoC] Added TF and PyTorch classification conversion cases

### PR DESCRIPTION
**Pull request by LupusSanctus**
GSoC project: "Write a tutorial about training a network in TF or PyTorch and then running it in OpenCV"

**Part 1**. Classification case.
* Provided dnn_model_runner module with the below-listed functions:
1. evaluation mode: a full pipeline for the evaluation of PyTorch and TF models (conversion, OpenCV Net retrievement, data fetch and preparation, evaluation)
2. test mode: integrated with samples/dnn/classification.py for inference execution
3. includes autonomous scripts with step by step TF/PyTorch models => OpenCV Net conversion and inference (described in tutorials)

* Provided tutorial series 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    @abstractmethod
+    def preprocess(self, img):
+        pass
+
+    def get_batch(self, imgs_names):
+        assert type(imgs_names) is list
+        batch = np.zeros((len(imgs_names), 3, self.frame_size, self.frame_size)).astype(np.float32)
+        for i in range(len(imgs_names)):
+            img_name = imgs_names[i]
+            # changed to os.path.join
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py:38]: @LupusSanctus let's remove the comment?

File: samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            img_file = os.path.join(self.imgs_dir, img_name)
+            assert os.path.exists(img_file)
+            img = cv2.imread(img_file, cv2.IMREAD_COLOR)
+            min_dim = min(img.shape[-3], img.shape[-2])
+            resize_ratio = self.frame_size / float(min_dim)
+            img = cv2.resize(img, (0, 0), fx=resize_ratio, fy=resize_ratio)
+            cols = img.shape[1]
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py:45]: I suggest `rows, cols, _ = img.shape`

File: samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            rows = img.shape[0]
+            y1 = round((rows - self.frame_size) / 2)
+            y2 = round(y1 + self.frame_size)
+            x1 = round((cols - self.frame_size) / 2)
+            x2 = round(x1 + self.frame_size)
+            img = img[y1:y2, x1:x2]
+            if self.bgr_to_rgb:
+                img = img[..., ::-1]
+            image_data = img[:, :, 0:3].transpose(2, 0, 1)
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py:54]: I suggest to split this code to several small functions like `center_crop` and `preprocess_image` to make it more readable.

File: samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            batch[i] = self.preprocess(image_data)
+        return batch
+
+
+class NormalizedValueFetch(DataFetch):
+    def __init__(self, imgs_dir, frame_size, bgr_to_rgb):
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py:60]: It would be good to declare the expected argument types in the signature.

File: samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        image_data /= 255.0
+        return image_data
+
+
+class MeanValueFetch(DataFetch):
+    def __init__(self, frame_size, imgs_dir, bgr_to_rgb):
+        self.imgs_dir = imgs_dir
+        self.frame_size = frame_size
+        self.mean_blob = np.ones((3, self.frame_size, self.frame_size)).astype(np.float32)
+        self.mean_blob *= 117
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/common/accuracy_eval.py:77]: What is 117? Looks like a magic number - let's avoid them.

File: samples/python/tutorial_code/dnn/dnn_conversion/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# Python 3.7.5
+torch==1.4.0
+numpy==1.18.4
+dataclasses==0.6
+tensorflow==2.1.0
+tensorflow-gpu==2.1.0
+# manual python binding
+opencv==4.3.0
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/requirements.txt:8]: I don't think we can do this. @asmorkalov please take a look.

File: samples/python/tutorial_code/dnn/dnn_conversion/tf/classification/py_to_py_resnet50.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    # Test the process of model retrieval
+    resnets = TFResNet50(
+        model_name="resnet50",
+        original_model=ResNet50(
+            include_top=True, weights="imagenet"
+        )
+    )
+    model_dict = resnets.get_prepared_models()
+    layers = model_dict["DNN resnet50"].getLayerNames()
+    print(layers)
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/tf/classification/py_to_py_resnet50.py:30]: Do we need this output?

File: samples/python/tutorial_code/dnn/dnn_conversion/tf/tf_model.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+class TFModelProcessor(Framework):
+    def __init__(self, prepared_model, model_name):
+        self._prepared_model = prepared_model
+        self._name = model_name
+
+    def get_output(self, input_blob):
+        assert len(input_blob.shape) == 4
+        batch_tf = input_blob.transpose(0, 2, 3, 1)
+        out = self._prepared_model(batch_tf)[..., 1:1001]
```
Comment (AnnaPetrovicheva) [inline samples/python/tutorial_code/dnn/dnn_conversion/tf/tf_model.py:63]: Would be better to change the indices to `1:self.num_classes + 1`.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+## Goals
+In this tutorial you will learn how to:
+* convert PyTorch classification models into ONNX format
+* run converted PyTorch model with OpenCV Python API
+* obtain an evaluation of the PyTorch and OpenCV DNN models
+
+We will explore the above-listed points by the example of the ResNet-50 architecture.
+
+## Introduction
+Let's briefly view the key concepts involved in the pipeline of PyTorch models transition with OpenCV API. The initial step in conversion of PyTorch models into [cv.dnn_Net](https://docs.opencv.org/4.3.0/db/d30/classcv_1_1dnn_1_1Net.html#a82eb4d60b3c396cb85c79d267516cf15)
```
Comment (alalek) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md:12]: > https://docs.opencv.org/4.3.0/

Please remove such links from the document.
Doxygen reference should be used instead

File: samples/dnn/common.py [hunk 10]
```diff
@@ -107,6 +108,6 @@ def findFile(filename):
                 pass
 
         print('File ' + filename + ' not found! Please specify a path to '
-              '/opencv_extra/testdata in OPENCV_DNN_TEST_DATA_PATH environment '
-              'variable or pass a full path to model.')
+                                   '/opencv_extra/testdata in OPENCV_DNN_TEST_DATA_PATH environment '
+                                   'variable or pass a full path to model.')
```
Comment (alalek) [inline samples/dnn/common.py:112]: Please avoid unnecessary changes

File: samples/dnn/common.py [hunk 11]
```diff
@@ -75,7 +76,7 @@ def add_preproc_args(zoo, parser, sample):
                  help='Preprocess input image by resizing to a specific width.')
     add_argument(zoo, parser, 'height', type=int,
                  help='Preprocess input image by resizing to a specific height.')
-    add_argument(zoo, parser, 'rgb', action='store_true',
+    add_argument(zoo, parser, 'rgb', type=bool, default=True,
```
Comment (alalek) [inline samples/dnn/common.py:79]: > default=True

Breaking change

**Review by alalek — COMMENTED** [review]
Need to clarify licensing terms of used images.

File: samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+torchvision>=0.6.1
+
+tensorflow>=2.1.0
+tensorflow-gpu>=2.1.0
+
+# manual python binding
+opencv>=4.4.0
+
+# nightly versions for segmentation experiments
+torch>=1.7.0.dev20200727
```
Comment (asmorkalov) [inline samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt:15]:
```
Double requirement given: torch>=1.7.0.dev20200727 (from -r ./requirements.txt (line 15)) (already in torch>=1.5.1 (from -r ./requirements.txt (line 5)), name='torch')
```

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+# get preprocessed image
+input_img = get_preprocessed_img("../data/ILSVRC2012_val_00000502.JPEG")
+
+# get ImageNet labels
+imagenet_labels = get_imagenet_labels("../data/dnn/classification_classes_ILSVRC2012.txt")
+
+# obtain OpenCV DNN predictions
+get_opencv_dnn_prediction(opencv_net, input_img, imagenet_labels)
+
+# obtainTF model predictions
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:60]: space is missing.

File: samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt (hunk 12 above)
Comment (asmorkalov) [reply samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt:15]: Is there any reason to use non-stable PyTorch version?

File: samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+torchvision>=0.8.0.dev20200727
```
Comment (asmorkalov) [inline samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt:16]: `dataclasses` module is missing for Python distros before 3.7.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
There is conflict in `requirements.txt`. Pip reports:
```
ERROR: Double requirement given: torch>=1.7.0.dev20200727 (from -r ./requirements.txt (line 15)) (already in torch>=1.5.1 (from -r ./requirements.txt (line 5)), name='torch')
```

**Comment by asmorkalov** [comment]
Please add bock with setup instructions. Let's expect that user already have python3 and pip installed.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
@LupusSanctus Sorry for confusion, please ignore the `requirements.txt` issue. I used old version of the code. Please remove OpenCV package from `requirements.txt` and add line to tutorial or just leave OpenCV without version. 4.5.x is not available with pip now.

**Comment by alalek** [comment]
Please rebase commits to resolve merge conflicts.

**Comment by LupusSanctus** [comment]
@asmorkalov, setup instructions were added in tutorial `Requirements`  section. Also `requirements.txt` was corrected.

**Comment by alalek** [comment]
Is this PR in WIP status?

**Review by alalek — COMMENTED** [review]
Well done!

Please take a look on minor comments below.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+    size=(224, 224),  # img target size
+    mean=mean,
+    swapRB=True,  # BGR -> RGB
+    crop=True  # center crop
+)
+print("Input blob shape: {}\n".format(input_blob.shape))
+```
+
+Please, pay attention at the preprocessing order in the cv2.dnn.blobFromImage function. Firstly, the mean value is subtracted and only then pixel values are multiplied by the defined scale.
+Therefore, to reproduce the image preprocessing pipeline from the TF [``mobilenet.preprocess_input``](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/applications/mobilenet.py#L443-L445) function, we multiply ``mean`` by ``127.5``.
```
[the remark this answers is not in the document]
Comment (ichisadashioko) [reply doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:172]: Here is the permalink - `https://github.com/tensorflow/tensorflow/blob/02032fb477e9417197132648ec81e75beee9063a/tensorflow/python/keras/applications/mobilenet.py#L443-L445`.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+
+## Requirements
+To be able to experiment with the below code you will need to install a set of libraries. We will use a virtual environment with python3.7+ for this:
+
+```console
+virtualenv -p /usr/bin/python3.7 <env_dir_path>
+source <env_dir_path>/bin/activate
+```
+
+For OpenCV-Python building from source, follow the corresponding instructions from the [OpenCV-Python tutorial section](https://docs.opencv.org/master/da/df6/tutorial_py_table_of_contents_setup.html).
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:29]: Please use Doxygen @ref here instead of direct link.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+
+Before you start the installation of the libraries, you can customize the [requirements.py](https://github.com/opencv/opencv/tree/master/samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt), excluding or including (for example, ``opencv-python``) some dependencies.
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:31]: `requirements.txt` in link.

File: samples/dnn/classification.py [hunk 18]
```diff
@@ -1,85 +1,112 @@
...
-
-    # Print predicted class.
-    label = '%s: %.4f' % (classes[classId] if classes else 'Class #%d' % classId, confidence)
-    cv.putText(frame, label, (0, 40), cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0))
-
-    cv.imshow(winName, frame)
+
+def get_args_parser(func_args):
+    backends = (cv.dnn.DNN_BACKEND_DEFAULT, cv.dnn.DNN_BACKEND_HALIDE, cv.dnn.DNN_BACKEND_INFERENCE_ENGINE,
+                cv.dnn.DNN_BACKEND_OPENCV)
```
Comment (asmorkalov) [inline samples/dnn/classification.py:10]: CuDNN?

File: samples/dnn/classification.py [hunk 19]
```diff
@@ -1,85 +1,112 @@
...
+        # Create a 4D blob from a frame.
+        inpWidth = args.width if args.width else frame.shape[1]
+        inpHeight = args.height if args.height else frame.shape[0]
+
+        if args.initial_width and args.initial_height:
+            frame = cv.resize(frame, (args.initial_width, args.initial_height))
+
+        blob = cv.dnn.blobFromImage(frame, args.scale, (inpWidth, inpHeight), args.mean, args.rgb, crop=args.crop)
+        if args.std:
+            blob[0] /= np.asarray(args.std, dtype=np.float32).reshape(3, 1, 1)
```
Comment (asmorkalov) [inline samples/dnn/classification.py:88]: Please add check to exclude devision by zero as in C++ version.

File: samples/dnn/dnn_model_runner/dnn_conversion/common/img_utils.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    if is_bgr_to_rgb:
+        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+    return img
+
+
+def get_pytorch_preprocess(img):
+    img = img.astype(np.float32)
+    img *= BASE_IMG_SCALE_FACTOR
+    img -= [0.485, 0.456, 0.406]
+    img /= [0.229, 0.224, 0.225]
```
Comment (asmorkalov) [inline samples/dnn/dnn_model_runner/dnn_conversion/common/img_utils.py:18]: The same values are in `default_preprocess_config.py`. Let's reuse the constants.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
@LupusSanctus Friendly reminder.

**Review by asmorkalov — APPROVED** [review]

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:
